File: Back/NodeJS/modules/bets/src/application/features/retrieve-bets/RetrieveBetsQueryHandler.ts
import { IRequestHandler } from '../../../../../shared/application/Request/IRequestHandler';
import { IRequest } from '../../../../../shared/application/Request/IRequest';
import { IUserContext } from '../../Abstractions/IUserContext';
export class RetrieveBetsQueryHandler implements IRequestHandler<RetrieveBetsQuery, RetrieveBetsResponse[]> {
    constructor(private readonly dataAccess: IRetrieveBetsDataAccess,
        private userContext: IUserContext){}
    Handle(request: RetrieveBetsQuery): Promise<RetrieveBetsResponse[]> {
        return this.dataAccess.RetrieveBetsAsync(this.userContext.GetUserId());
    }
    GetRequestType(): string {
        return RetrieveBetsQuery.name;
    }
}

export class RetrieveBetsQuery extends IRequest<RetrieveBetsResponse[]> {
    constructor(){
        super();
    }
    Name: string = RetrieveBetsQuery.name;
}

export interface RetrieveBetsResponse {
    Id: string;
    Description: string;
    Coins: number;
    EndDate: Date;
    BookieId: string;
    BookieName: string;
    MaxAnswerDate: Date;
    Gamblers: GamblerDto[]
}

export interface IRetrieveBetsDataAccess {
    RetrieveBetsAsync(userId: string): Promise<RetrieveBetsResponse[]>;
}

export interface GamblerDto {
    Id: string;
    Name: string;
    HasAccepted?: boolean;
}

File: Back/.Net/Modules/Bets/BetFriends.Bets.Application/Features/RetrieveBets/RetrieveBetsQueryHandler.cs
﻿using BetFriends.Bets.Application.Abstractions;
using BetFriends.Shared.Application.Abstractions.Messaging;

namespace BetFriends.Bets.Application.Features.RetrieveBets;

public class RetrieveBetsQueryHandler(IRetrieveBetsDataAccess retrieveBetsDataAccess,
                                      IUserContext userContext) : IQueryHandler<RetrieveBetsQuery, IEnumerable<RetrieveBetsResponse>>
{
    private readonly IRetrieveBetsDataAccess retrieveBetsDataAccess = retrieveBetsDataAccess;
    private readonly IUserContext userContext = userContext;

    public Task<IEnumerable<RetrieveBetsResponse>> Handle(RetrieveBetsQuery request, CancellationToken cancellationToken)
    {
        return retrieveBetsDataAccess.GetAsync(userContext.UserId);
    }
}


public record RetrieveBetsQuery : IQuery<IEnumerable<RetrieveBetsResponse>>;
public record RetrieveBetsResponse(Guid Id,
                                   string Description,
                                   DateTime EndDate,
                                   int Coins,
                                   Guid BettorId,
                                   string BettorName,
                                   DateTime MaxAnswerDate,
                                   IEnumerable<GamblerDto> Gamblers);

public record GamblerDto(Guid Id, string Name, bool? HasAccepted);

public interface IRetrieveBetsDataAccess
{
    Task<IEnumerable<RetrieveBetsResponse>> GetAsync(Guid memberId);
}